### DISCUSSION TITLE
Hardware recommendation

### DISCUSSION THREAD

**Discussion by aronymous** [blakeblackshear/frigate#1758] [General]
I am looking to move all my cameras over to frigate, and looking for hardware recommendations for coral and gpu. 

My plan is to use docker (k8s) on a bare metal linux os. This machine has PCIe ports free. 

Cameras:
5 various types of camera (all support rtsp)
Resolutions vary from 720p up to 4k. 

More specs:
Intel(R) Xeon(R) CPU           L5640  24 cores total. Older cpu but lots of old power. 
32gb of ram. 

I am thinking of getting Coral Dual https://coral.ai/products/m2-accelerator-dual-edgetpu/ but not sure if it will give let me support more cameras in the future. 
I would like to be able to detect and record on the full resolution and try to use 
Thanks

**Comment by haggercody** [comment]
i recently got a coral and it dropped my inference time from 100ms to about 30ms for 3 cameras.  I have  a zwave powerstrip that reports energy consumption as well, and It didn't change much at all--if anything went up a bit.  The biggest perf improvement I got was enabling quicksync for the hwaccel.

**Comment by tve** [comment]
You may want to do some searches WRT the dual edgetpu board. Lots of people are having trouble with it. Some PCI slots don't support the 2x thing, some drivers bark... Not saying it can't work...

**Reply by aronymous** [reply]
Thanks. Maybe i will stick with single board to start.

**Reply by pdecat** [reply]
This [m.2 E Key PCIe 2x to m.2 B+M Key PCIe 1x adapter](https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter#m2-bm-version) coupled with a [generic m.2 B+M Key to PCIe 1x adapter](https://www.amazon.com/Cablecc-Carte-Adaptateur-M-Key-Express/dp/B00S6B6MZW/) will make the [Coral M.2 Accelerator with Dual Edge TPU](https://coral.ai/products/m2-accelerator-dual-edgetpu/) work in a desktop PCIe 1x slot.

I've been using it for almost two months without issues: https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter/issues/3#issuecomment-885755446

A direct [m.2 E Key PCIe 2x to PCIe 1x](https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter#low-profile-pcie-x1-version) should be [available for purchase in October](https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter/issues/4#issuecomment-912628771).

**Reply by DrSpaldo** [reply]
> This [m.2 E Key PCIe 2x to m.2 B+M Key PCIe 1x adapter](https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter#m2-bm-version) coupled with a [generic m.2 B+M Key to PCIe 1x adapter](https://www.amazon.com/Cablecc-Carte-Adaptateur-M-Key-Express/dp/B00S6B6MZW/) will make the [Coral M.2 Accelerator with Dual Edge TPU](https://coral.ai/products/m2-accelerator-dual-edgetpu/) work in a desktop PCIe 1x slot.
> 
> I've been using it for almost two months without issues: [magic-blue-smoke/Dual-Edge-TPU-Adapter#3 (comment)](https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter/issues/3#issuecomment-885755446)
> 
> A direct [m.2 E Key PCIe 2x to PCIe 1x](https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter#low-profile-pcie-x1-version) should be [available for purchase in October](https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter/issues/4#issuecomment-912628771).

Where do you get that https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter#m2-bm-version you mentioned from?

**Reply by pdecat** [reply]
I got a prototype for testing.

Production for m.2 B+M adapters should come after the desktop PCIe ones: https://github.com/magic-blue-smoke/Dual-Edge-TPU-Adapter/issues/3#issuecomment-893755012

**Comment by tubalainen** [comment]
Hope you can exhaust all the heat the Xeons will generate. As well as placing it somewhere where it´s not causing any issues with the fan noise. Got my self an old HP rack server and had a stab at a similar setup, had to drop it and ive got myself some core i5-7500T machines instead. The Xeon heated up my man cave shed to 30 degrees during winter and I could not work at my desk without hearing protection. I do miss the capacity and freedom of tinkering it brought though ... 

On the GPU side, a Nvidia 1060 ti and a Google Coral will offload the CPU a lot when running Frigate.

**Reply by Ysbrand** [reply]
Just wondering, what is the power usage of the I5-7500T setup?

**Reply by tubalainen** [reply]
> Just wondering, what is the power usage of the I5-7500T setup?

My Frigate server peaks at ~38 watt and "idles" around 30 watts under a 50% CPU load.

I ended up getting my a used HP Elitedesk 800 mini G3 with the i5-7500T CPU, 8 gigs of mem to run frigate. 250 EUR (ish) off ebay.com.

TDP is 35W on the CPU.
https://ark.intel.com/content/www/us/en/ark/products/97121/intel-core-i5-7500t-processor-6m-cache-up-to-3-30-ghz.html

**Reply by aronymous** [reply]
Yes the server is in my basement. perfectly cool enough and far enough away not to have to hear it. 

I might be able to get a old 1030 for free so it will be a good test of decoding. 

Thanks for the info!

**Comment by pdecat** [comment]
I believe you'll need a custom build of tflite to support that CPU: https://ark.intel.com/content/www/fr/fr/ark/products/47926/intel-xeon-processor-l5640-12m-cache-2-26-ghz-5-86-gt-s-intel-qpi.html

See https://github.com/blakeblackshear/frigate/issues/49
and https://github.com/blakeblackshear/frigate-hass-addons/pull/12

**Comment by antonwantstosleep** [comment]
I've got a rack server with an Intel Xeon Silver 4214R CPU @ 2.40GHz 12 cores installed in Supermicro X11SPi-TF and no GPU. It runs Proxmox with several VM`s for my office (freeipa domain controller, samba file server, freeswitch telephony etc.). One of them is a supervised Home Assistant for automations (2 cores, 4 GB RAM).

My goal is to record 24/7 and monitor about 11+10+3=24 cameras 1920x1080 H264 split into 3 groups (buildings). I don't want to buy ready NVR`s because I want to ~~have a sex~~ control everything.

I tried to use Frigate as a HA addon. But my tests are:
- 2 cameras with detect only enabled - 80% of CPU
- 5 cameras with everything disabled (detect, record, rtmp) - 80% of CPU

I didn't enable hardware acceleration because as I understand right my CPU has no graphics (and there is no /dev/dri/renderD128 in /dev). Correct me if I am wrong please.

So my question is: what should I do? :) I think my options are:
- buy a Coral device for Xeon
- buy a Coral device and a GPU (which one?) for Xeon
- buy 3 RPi 4 8GB and 3 Coral USB devices (and 3 USB HDD 4TB) in order to make a distributed system (each bulding would have it`s own NVR.. but is RPi able to handle 11 cameras?)
- kill myself

Thank you!

**Reply by blakeblackshear** [reply]
The Coral does not help with decoding the video stream, so it won't reduce the load you see with detect off. What is the frame rate for your cameras? If you have the ability to lower it on the cameras, that will help quite a bit.

If you do end up building a distributed system, I don't recommend the Pi. It only has so much RAM you can allocate for the GPU, so you will hit a ceiling for the number of cameras. I would look for an intel x86 mini PC that supports quicksync.

**Reply by antonwantstosleep** [reply]
It`s 25 FPS and I think I can lower it but my old 8-channel DAHUA NVR 4108-P is ok with recording them. Sadly I can't find any information about it's CPU/GPU etc. I still don't understand - do I need the GPU?

**Reply by blakeblackshear** [reply]
Recording isn't the same as decoding. I'm guessing your old NVR is simply recording. Frigate can record high resolution and high frame rate cameras with the record role easily with barely any CPU usage. What takes a lot of CPU/GPU is decoding the video streams for analysis. That is why it is recommended to do analysis on a lower resolution stream at 5fps and then store the higher resolution stream. You can reduce CPU usage by either using a lower resolution stream and lower fps or getting a GPU with that Xeon. Either way, it is still recommended to use the lower resolution/fps stream first.
